lib/AArch64Base.ml

```diff
@@ -418,27 +418,51 @@ let barrier_compare = compare
 module IC = struct
   type funct = I
   let pp_funct = function I -> "I"
+  let fold_funct f k = f I k
 
   type typ = ALL | VA
   let pp_typ = function | ALL -> "ALL" | VA -> "VA"
+  let fold_typ f k = k |> f ALL |> f VA
 
   type point = U
   let pp_point = function U -> "U"
+  let fold_point f k = f U k
 
   type domain = IS | NO
   let pp_domain = function IS -> "IS" | NO -> ""
+  let fold_domain f k = k |> f IS |> f NO
 
   type op = { funct:funct; typ:typ; point:point; domain:domain; }
   let ivau = { funct=I; typ=VA; point=U; domain=NO; }
   let iallu = { funct=I; typ=ALL; point=U; domain=NO; }
 
+  let fold_op f k =
+    fold_funct
+      (fun funct k ->
+        fold_typ
+          (fun typ k ->
+            fold_point
+              (fun point k ->
+                fold_domain
+                  (fun domain k -> f {funct; typ; point; domain; } k)
+                  k)
+              k)
+          k)
+      k
+
   let pp_op op =
     pp_funct op.funct ^
     pp_typ op.typ ^
     pp_point op.point ^
     pp_domain op.domain
 
+  let do_pp tag op = "IC" ^ tag ^ (pp_op op)
+  let pp = do_pp " "
+  let pp_dot = do_pp "."
+
   let all op = match op.typ with | VA -> false | ALL -> true
+
+  let compare ic1 ic2 = compare ic1 ic2 = 0
```

Comment:
Usually `compare` returns an `int`, and not a `bool`. Should this be named `equal`?

Reply:
And is there a reason why you are not using `=`? As in:
```suggestion
  let compare ic1 ic2 = ic1 = ic2
```

Reply:
I thought of naming it `equal` but hesitated, other modules seem to use `compare`, in the same way. But since you spotted this as well, I'm going to change it to `equal`. But now I am thinking do we need to define `equal`, should we just use `=`?

Reply:
@HadrienRenaud let me know if using `=` seems ok. I've pushed the update and if you're happy I am going to squash all fixups.

Reply:
From my understanding:
It is a bit dangerous to use `=` on recursive types, as you are using something that is not in constant time, which is why we implement equal on complex types. Here everything is quite simple so I think we could do that.
However, if in the future we change this type to something more complicated (such as having Strings or such), we might not change those quite implicit `=`
So this solution is good for now, but the solution I think is the best as a best would be to have under the definition something more explicit like
```
let equal: op -> op -> bool  = ( = )
```

Anyway, if you think I'm just nitpicking, do squash.

Reply:
No you're not nitpicking :)

I've changed to `equal` and squashed all patches together. Thanks again for the review!